### PR TITLE
Extended account registration gas amount

### PR DESCRIPTION
**Pull request by platfowner**
Change summary:
- Apply account registration gas amount for nonce and timestamp in addition to the balance 
- Guard extended account registration gas amount with extend_account_registration_gas_amount timer flag
- Make nonce available in transaction execution (not necessary for this PR)
- Invalidate multi-set transactions with empty op_list
- Add more test cases
- Fix some typos in package.json 

Related issues:
- https://github.com/ainblockchain/ain-blockchain/issues/1016

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -34,8 +34,8 @@
...
     "test_unit_event_handler": "BLOCKCHAIN_CONFIGS_DIR=blockchain-configs/1-node ENABLE_REST_FUNCTION_CALL=true UNSAFE_PRIVATE_KEY=b22c95ffc4a5c096f7d7d0487ba963ce6ac945bdc91c79b64ce209de289bec96 ./node_modules/mocha/bin/mocha --timeout 160000 test/unit/event-handler.test.js",
-    "test_unit_functions": "BLOCKCHAIN_CONFIGS_DIR=blockchain-configs/1-nodeE NABLE_REST_FUNCTION_CALL=true UNSAFE_PRIVATE_KEY=b22c95ffc4a5c096f7d7d0487ba963ce6ac945bdc91c79b64ce209de289bec96 ./node_modules/mocha/bin/mocha --timeout 160000 test/unit/functions.test.js",
-    "test_unit_network_util": "BLOCKCHAIN_CONFIGS_DIR=blockchain-configs/1-nodeE NABLE_REST_FUNCTION_CALL=true UNSAFE_PRIVATE_KEY=b22c95ffc4a5c096f7d7d0487ba963ce6ac945bdc91c79b64ce209de289bec96 ./node_modules/mocha/bin/mocha --timeout 160000 test/unit/network-util.test.js",
+    "test_unit_functions": "BLOCKCHAIN_CONFIGS_DIR=blockchain-configs/1-node ENABLE_REST_FUNCTION_CALL=true UNSAFE_PRIVATE_KEY=b22c95ffc4a5c096f7d7d0487ba963ce6ac945bdc91c79b64ce209de289bec96 ./node_modules/mocha/bin/mocha --timeout 160000 test/unit/functions.test.js",
```
Comment (liayoo) [inline package.json:37]: 👀 thanks for fixing them!

File: common/path-util.js [hunk 2]
```diff
@@ -7,16 +7,20 @@ const RuleUtil = require('../db/rule-util');
...
+  }
+
   static getAccountBalancePath(address) {
-    return CommonUtil.formatPath([PredefinedDbPaths.ACCOUNTS, address, PredefinedDbPaths.BALANCE]);
+    return `${PathUtil.getAccountPath(address)}/${PredefinedDbPaths.BALANCE}`;
   }
 
   static getAccountNoncePath(address) {
-    return CommonUtil.formatPath([PredefinedDbPaths.ACCOUNTS, address, PredefinedDbPaths.ACCOUNTS_NONCE]);
+    return `${PathUtil.getAccountPath(address)}/${PredefinedDbPaths.ACCOUNTS_NONCE}`;
```
Comment (liayoo) [inline common/path-util.js:19]: ditto

File: common/path-util.js [hunk 3]
```diff
@@ -7,16 +7,20 @@ const RuleUtil = require('../db/rule-util');
...
   }
 
   static getAccountTimestampPath(address) {
-    return CommonUtil.formatPath([PredefinedDbPaths.ACCOUNTS, address, PredefinedDbPaths.ACCOUNTS_TIMESTAMP]);
+    return `${PathUtil.getAccountPath(address)}/${PredefinedDbPaths.ACCOUNTS_TIMESTAMP}`;
```
Comment (liayoo) [inline common/path-util.js:23]: ditto

File: db/index.js [hunk 4]
```diff
@@ -1316,13 +1328,29 @@ class DB {
...
+          this.executeMultiSetOperation(op.op_list, auth, nonce, timestamp, tx, blockNumber, blockTime));
     } else {
       Object.assign(
-          result, this.executeSingleSetOperation(op, auth, timestamp, tx, blockNumber, blockTime));
+          result, this.executeSingleSetOperation(op, auth, nonce, timestamp, tx, blockNumber, blockTime));
+    }
+    if (isEnabledTimerFlag('extend_account_registration_gas_amount', blockNumber)) {
+      // Apply account registration gas amount for nonce and timestamp.
+      const isNonExistingAccount = this.checkIfNonExistingAccount(tx, auth);
+      if (wasNonExistingAccount && isNonExistingAccount && nonce !== -1) {
```
Comment (liayoo) [inline db/index.js:1346]: 🙌

File: db/index.js [hunk 5]
```diff
@@ -1265,7 +1266,7 @@ class DB {
     return result;
   }
 
-  executeMultiSetOperation(opList, auth, timestamp, tx, blockNumber, blockTime) {
+  executeMultiSetOperation(opList, auth, nonce, timestamp, tx, blockNumber, blockTime) {
```
Comment (liayoo) [inline db/index.js:1269]: question) are the nonces added to executeMultiSetOperation() and executeSingleSetOperation() in order to match the function signature of executeOperation()?

File: common/path-util.js [hunk 6]
```diff
@@ -7,16 +7,20 @@ const RuleUtil = require('../db/rule-util');
 const ruleUtil = new RuleUtil();
 
 class PathUtil {
+  static getAccountPath(address) {
+    return CommonUtil.formatPath([PredefinedDbPaths.ACCOUNTS, address]);
+  }
+
   static getAccountBalancePath(address) {
-    return CommonUtil.formatPath([PredefinedDbPaths.ACCOUNTS, address, PredefinedDbPaths.BALANCE]);
+    return `${PathUtil.getAccountPath(address)}/${PredefinedDbPaths.BALANCE}`;
```
[the remark this answers is not in the document]
Comment (platfowner) [reply common/path-util.js:15]: At first I implemented so, but realized it's not efficient as the input paths are converted to arrays and then converted to a string. So, in the case the input format is clear I think just using the string concatenation is faster and simpler.

File: db/index.js (hunk 5 above)
Comment (platfowner) [reply db/index.js:1269]: Ah, at first I added it to use it inside the function call but realized that the account registration gas amount from the nonce or timestamp needs to be computed in executeOperation() with the following reasons:
- Nonce / timestamp update is done in executeOperation() per transaction
- One transaction can have multiple operations (e.g. SET operation)

As a result the change is not necessary in this PR, but worth to keep the change for further use cases in the future?

**Comment by platfowner** [comment]
Let me leave some quick answers to the questions and suggestions. PTAL~ @liayoo

File: common/path-util.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (liayoo) [reply common/path-util.js:15]: I see! Sounds good to me. Thanks for the explanation.

File: db/index.js (hunk 5 above)
Comment (liayoo) [reply db/index.js:1269]: Right. Makes sense & I agree that we can keep the nonce args for future use.

File: test/integration/sharding.test.js [hunk 7]
```diff
@@ -491,7 +491,7 @@ describe('Sharding', () => {
           [server2Addr]: true,
           [server3Addr]: true,
         }
-      });
+      }, 200);
```
Comment (liayoo) [inline test/integration/sharding.test.js:494]: 👍

File: tx-pool/transaction.js [hunk 8]
```diff
@@ -242,6 +242,10 @@ class Transaction {
       logger.info(`Transaction body is in a non-standard format:\n${diffLines}\n`);
       return false;
     }
+    if (sanitized.operation.type === WriteDbOperations.SET &&
```
Comment (liayoo) [inline tx-pool/transaction.js:245]: Huh, I thought we were already checking this somewhere, but I guess not.. Thanks for adding it!

**Review by liayoo — APPROVED** [review]
LGTM, thanks!

**Comment by platfowner** [comment]
Thanks for the review, now merging.. @liayoo

**Pull request merged**